### ISSUE TITLE
Heap memory area corruption error when using "--in-memory flag" with multi-threaded application

### ISSUE THREAD

**Issue by shinji-TH** [emscripten-core/emscripten#19091]
@sbc100 
In the documentation [Module Splitting](https://emscripten.org/docs/optimizing/Module-Splitting.html?highlight=split_module), it is mention that when doing module splitting with multi-threaded, the flag "--in-memory" and "-sGLOBALBASE=N" is required in order to capture all profile data from all the threads. 

I tested adding the flag "--in-memory"  under the argument for generating the instrument build [link](https://github.com/emscripten-core/emscripten/blob/de7cbc56c3ac38b83cc3e8bbc17ac5fbb6b3bbd7/emcc.py#L782) with the following flags
`-O0 --s ASSERTIONS=1 -s SAFE_HEAP=0 -pthread -s PROXY_TO_PTHREAD -s SPLIT_MODULE=1`

The build resulted in the following error:
![image](https://user-images.githubusercontent.com/33003856/228448149-8f4a6d8c-abfc-4033-8de8-5a48ed9d396a.png)

The error was due to a failure in heap check [runtime_stack_check.js](https://github.com/emscripten-core/emscripten/blob/de7cbc56c3ac38b83cc3e8bbc17ac5fbb6b3bbd7/src/runtime_stack_check.js#L53)
` if (HEAPU32[0] !== 0x63736d65 /* 'emsc' */) abort('Runtime error: The application has corrupted its heap memory area (address zero)!');`
Printing the HEAPU32[0] values in initRunTime shows that the values changes after __RELOC_FUNCS__, which equates to Module["asm"]["__wasm_apply_data_relocs"]
![image](https://user-images.githubusercontent.com/33003856/228450198-cb751650-84d4-4802-9a18-ed24b949943a.png)
Looking into the implementation of "__wasm_apply_data_relocs" in wat for instrument build, the function "__wasm_apply_data_relocs" is being marked via in the location  i32 atomic.store8 offset=2
>   (func $__wasm_apply_data_relocs (type 5)
    i32.const 0
    i32.const 1
    i32.atomic.store8 offset=2
    block  ;; label = @1
      i32.const 21844
      global.get $__memory_base
      i32.add
      global.get $__memory_base
      i32.const 6640

In the function writeStackCookie() in runtime_stack_check.js, `HEAPU32[0] = 0x63736d65; /* 'emsc' */`
0x63736d65 = 1668509029
The instrumentated build doing a store8 of value 1 by offsetting 2
0x63016d65 = 1661037925
Which equals to the value after __RELOC_FUNCS__, 
This means that the flag "--in-memory" is using the start of heap to store the profile data, which resulted in overwriting the values in HEAPU32[0]. 

Building the app with `-s SAFE_HEAP=1`, would result in no errors or corruption of heap. 

Hence wish to check if this behaviour is intended and if there is any potential issue with the usage of "--in-memory" flag using the start of HEAPU32 for profile data. 

Tested with emsdk 3.1.34

**Comment by sbc100** [comment]
It looks like you are using dynamic linking (e.g. `-sMAIN_MODULE`) with module splitting.. is that correct?   Did you include all the link flags, because I don't see `-sMAIN_MODULE` or `-sRELOCATABLE` listed.

@tlively do we support module splitting in combination with dynamic linking?

**Comment by tlively** [comment]
Yes, the combination of dynamic linking and module splitting is supported and tested. The SAFE_HEAP checker is not aware that profiling data will be deliberately written to address zero, though, so it should probably be disabled for this application.

**Comment by shinji-TH** [comment]
Yes it is compile with dynamic linking, with module splitting. Here is the full list of flags that I am using for the main module
`    -O0
    --profiling
    --profiling-funcs
    -s ASSERTIONS=1
    -s SAFE_HEAP=0
    -s DEMANGLE_SUPPORT=1
    -lnodefs.js
    -std=c++20
    -D_LIBCPP_ENABLE_CXX17_REMOVED_AUTO_PTR
    -s INCLUDE_FULL_LIBRARY=1
    -s ERROR_ON_UNDEFINED_SYMBOLS=0
    -fwasm-exceptions
    -fstrict-aliasing
    -fPIC
    -s WASM_BIGINT=1
    -pthread
    -s PTHREAD_POOL_SIZE=2
    -s PROXY_TO_PTHREAD
    -s FETCH=1
    -s FETCH_SUPPORT_INDEXEDDB=0
    -Wno-deprecated-builtins
    -s WASM=1
    -s ALLOW_MEMORY_GROWTH=1
    -s EVAL_CTORS=0
    -s MAIN_MODULE=2
    -s INCLUDE_FULL_LIBRARY=1
    -s AUTOLOAD_DYLIBS=0
    -s NO_EXIT_RUNTIME=0
    -s INLINING_LIMIT=1
    -s FORCE_FILESYSTEM=1
    --emit-symbol-map
    -s LEGACY_GL_EMULATION=0
    -s INITIAL_MEMORY=1000MB
    -s ENVIRONMENT=web,worker
    -s USE_SDL=0
    -s EXPORT_EXCEPTION_HANDLING_HELPERS=1
    -s STRICT=1
    -s ALLOW_UNIMPLEMENTED_SYSCALLS=1
    -s SPLIT_MODULE=1
    -s USE_ICU=1`

**Comment by shinji-TH** [comment]
@tlively, does that mean that SAFE_HEAP should also reflect error when profiling data is written to address zero?
If so, is it safe to use the flag?

**Comment by tlively** [comment]
No, I don't think it's safe to use SAFE_HEAP in combination with SPLIT_MODULE (when using --in-memory profiling). Hopefully this can be fixed by using a separate memory to store the profile later this year.
